scripts/rollup/build.js

```diff
@@ -364,160 +362,125 @@ function createBundle(bundle, bundleType) {
+    fs.mkdirSync(join('build', 'dist'));
 
-  const tasks = [
-    Packaging.createFacebookWWWBuild,
-    Packaging.createReactNativeBuild,
-    Packaging.createReactNativeRTBuild,
-    Packaging.createReactNativeCSBuild,
-  ];
-  for (const bundle of Bundles.bundles) {
-    tasks.push(
-      () => createBundle(bundle, UMD_DEV),
-      () => createBundle(bundle, UMD_PROD),
-      () => createBundle(bundle, NODE_DEV),
-      () => createBundle(bundle, NODE_PROD),
-      () => createBundle(bundle, FB_DEV),
-      () => createBundle(bundle, FB_PROD),
-      () => createBundle(bundle, RN_DEV),
-      () => createBundle(bundle, RN_PROD)
-    );
-  }
-  if (syncFbsource) {
-    tasks.push(() =>
-      syncReactNative(join('build', 'react-native'), syncFbsource)
-    );
-    tasks.push(() =>
-      syncReactNativeRT(join('build', 'react-rt'), syncFbsource)
-    );
-    tasks.push(() =>
-      syncReactNativeCS(join('build', 'react-cs'), syncFbsource)
-    );
-  } else if (syncWww) {
-    tasks.push(() => syncReactDom(join('build', 'facebook-www'), syncWww));
-  }
-  // rather than run concurrently, opt to run them serially
-  // this helps improve console/warning/error output
-  // and fixes a bunch of IO failures that sometimes occurred
-  return runWaterfall(tasks)
-    .then(() => {
-      // output the results
-      console.log(Stats.printResults());
-      // save the results for next run
-      Stats.saveResults();
-      if (shouldExtractErrors) {
-        console.warn(
-          '\nWarning: this build was created with --extract-errors enabled.\n' +
-            'this will result in extremely slow builds and should only be\n' +
-            'used when the error map needs to be rebuilt.\n'
-        );
-      }
-    })
-    .catch(err => {
-      console.error(err);
-      process.exit(1);
-    });
-});
+    await Packaging.createFacebookWWWBuild();
+    await Packaging.createReactNativeBuild();
+    await Packaging.createReactNativeRTBuild();
+    await Packaging.createReactNativeCSBuild();
 
-function runWaterfall(promiseFactories) {
-  if (promiseFactories.length === 0) {
-    return Promise.resolve();
-  }
+    // Run them serially for better console output
+    // and to avoid any potential race conditions.
+    for (const bundle of Bundles.bundles) {
+      await createBundle(bundle, UMD_DEV);
+      await createBundle(bundle, UMD_PROD);
+      await createBundle(bundle, NODE_DEV);
+      await createBundle(bundle, NODE_PROD);
+      await createBundle(bundle, FB_DEV);
+      await createBundle(bundle, FB_PROD);
+      await createBundle(bundle, RN_DEV);
+      await createBundle(bundle, RN_PROD);
+    }
 
-  const head = promiseFactories[0];
-  const tail = promiseFactories.slice(1);
+    if (syncFbsource) {
+      await syncReactNative(join('build', 'react-native'), syncFbsource);
+      await syncReactNativeRT(join('build', 'react-rt'), syncFbsource);
+      await syncReactNativeCS(join('build', 'react-cs'), syncFbsource);
```

Comment:
nit: We could wrap these in a `Promise.all` (not that it would make a big diff, should be fast either way)

Reply:
The build script originally did exactly this but it made the output unreadable (the different bundle errors/warnings were spat out over each other) and there win wasn't actually that of a win in performance as the work is mostly CPU intensive, not IO intensive.

Reply:
That's for the create-bundle steps (below). The sync steps here just copy files. They should be okay to do in parallel. (They don't even log anything, so they wouldn't cause unreadable output.)

But, it wouldn't make a big diff in terms of performance since they're just copying a few files. So it's a "nit" 😄